Source/ObjectiveDropboxOfficial/Shared/Handwritten/Networking/DBGlobalErrorResponseHandler.m

```diff
@@ -0,0 +1,171 @@
+
+    if ([s_routeErrorToResponseBlock objectForKey:errorClass]) {
+      [queueToUse addOperationWithBlock:^{
+        [s_lock lock];
+        if (s_routeErrorToResponseBlock[errorClass]) {
+          s_routeErrorToResponseBlock[errorClass](routeError, error);
+        }
+        [s_lock unlock];
+      }];
+    }
+
+    for (Class instanceClass in instanceVariablesClassesToValues) {
+      if ([s_routeErrorToResponseBlock objectForKey:instanceClass]) {
+
+        [queueToUse addOperationWithBlock:^{
+          [s_lock lock];
+          if (s_routeErrorToResponseBlock[instanceClass]) {
+            s_routeErrorToResponseBlock[instanceClass](instanceVariablesClassesToValues[instanceClass], error);
+          }
+          [s_lock unlock];
+        }];
+      }
+    }
+    [s_lock unlock];
+  }
+
+  // execute network error block
+  if (error) {
+    [s_lock lock];
+    if (s_networkErrorResponseBlock) {
+      NSOperationQueue *queueToUse = s_networkErrorQueue;
+
+      [queueToUse addOperationWithBlock:^{
+        [s_lock lock];
+        if (s_networkErrorResponseBlock) {
+          s_networkErrorResponseBlock(error);
+        }
+        [s_lock unlock];
+      }];
+    }
+    [s_lock unlock];
+  }
+}
+
++ (NSDictionary<Class, id> *)instanceVariablesDataFromRouteError:(id)routeError {
+  Class errorClass = [routeError class];
+
+  NSMutableDictionary<Class, id> *result = [NSMutableDictionary new];
+
+  // from http://stackoverflow.com/questions/16861204/property-type-or-class-using-reflection
+  unsigned int count;
+  objc_property_t *props = class_copyPropertyList([errorClass class], &count);
+
+  for (int i = 0; i < count; i++) {
+    objc_property_t property = props[i];
+    const char *name = property_getName(property);
+    NSString *propertyName = [NSString stringWithCString:name encoding:NSUTF8StringEncoding];
+
+    const char *type = property_getAttributes(property);
+
+    NSString *typeString = [NSString stringWithUTF8String:type];
+    NSArray *attributes = [typeString componentsSeparatedByString:@","];
+    NSString *typeAttribute = [attributes objectAtIndex:0];
+    NSString *propertyType = [typeAttribute substringFromIndex:1];
+
+    if ([typeAttribute hasPrefix:@"T@"]) {
+      NSString *typeClassName =
+          [typeAttribute substringWithRange:NSMakeRange(3, [typeAttribute length] - 4)]; // turns @"NSDate" into NSDate
+      Class typeClass = NSClassFromString(typeClassName);
+      if (typeClass != nil && typeClass != [NSString class]) {
+        // we try / catch here because for union types, a runtime exception is thrown
+        // when a value type that doesn't correspond to the tag state is accessed.
+        @try {
+          id object = [routeError valueForKey:propertyName];
+          result[typeClass] = object;
+          NSDictionary<Class, id> *additionalData = [[self class] instanceVariablesDataFromRouteError:object];
+          [result addEntriesFromDictionary:additionalData];
+        } @catch (NSException *) {
```

Comment:
Have you tried measuring the impact of this? Probably negligible, but this seems like a lot of exceptions, and I don't think exceptions are particularly well-optimized in Obj-C.